src/rust/lib/graph-editor/src/component/visualization/container.rs

```diff
@@ -0,0 +1,174 @@
+//! This module defines the `Container` struct and related functionality.
+
+use crate::prelude::*;
+
+use crate::frp;
+use crate::visualization::*;
+
+use ensogl::display::traits::*;
+use ensogl::display;
+
+
+
+// ===========
+// === FRP ===
+// ===========
+
+/// Event system of the `Container`.
+#[derive(Clone,CloneRef,Debug)]
+#[allow(missing_docs)]
+pub struct ContainerFrp {
+    pub network           : frp::Network,
+    pub set_visibility    : frp::Source<bool>,
+    pub toggle_visibility : frp::Source,
+    pub set_visualization : frp::Source<Option<Visualization>>,
+    pub set_data          : frp::Source<Option<Data>>,
+}
+
+impl Default for ContainerFrp {
+    fn default() -> Self {
+        frp::new_network! { visualization_events
+            def set_visibility    = source::<bool>                  ();
+            def toggle_visibility = source::<()>                    ();
+            def set_visualization = source::<Option<Visualization>> ();
+            def set_data          = source::<Option<Data>>          ();
+        };
+        let network = visualization_events;
+        Self {network,set_visibility,set_visualization,toggle_visibility,set_data }
+    }
+}
+
+
+
+// ================================
+// === Visualizations Container ===
+// ================================
+
+/// Container that wraps a `Visualization` for rendering and interaction in the GUI.
+///
+/// The API to interact with the visualisation is exposed through the `ContainerFrp`.
+#[derive(Clone,CloneRef,Debug,Shrinkwrap)]
+#[allow(missing_docs)]
+pub struct Container {
+    // The internals are split into two structs: `ContainerData` and `ContainerFrp`. The
+    // `ContainerData` contains the actual data and logic for the `Container`. The `ContainerFrp`
+    // contains the FRP api and network. This split is required to avoid creating cycles in the FRP
+    // network: the FRP network holds `Rc`s to the `ContainerData` and thus must not live in the
+    // same struct.
+
+    #[shrinkwrap(main_field)]
+        data : Rc<ContainerData>,
+    pub frp  : ContainerFrp,
+}
+
+/// Internal data of a `Container`.
+#[derive(Debug,Clone)]
+#[allow(missing_docs)]
+pub struct ContainerData {
+    logger         : Logger,
+    display_object : display::object::Instance,
+    size           : Cell<Vector2<f32>>,
+    visualization  : RefCell<Option<Visualization>>,
+}
+
+impl ContainerData {
+    /// Set whether the visualisation should be visible or not.
+    pub fn set_visibility(&self, is_visible:bool) {
+        if let Some(vis) = self.visualization.borrow().as_ref() {
+            if is_visible {
+                vis.display_object().set_parent(&self.display_object);
+            } else {
+                vis.display_object().unset_parent();
```

Comment:
why are you using explicit calls to `display_object()`? just skip it. Here and in other places.